File: local/event.py


class Time:

    def __init__(self, hour, minute):
        self.hour = hour
        self.minute = minute

    def __eq__(self, other):
        return self.hour == other.hour and self.minute == other.minute

    def __str__(self):
        hour = str(self.hour)
        minute = str(self.minute)

        if self.hour < 10:
            hour = "0" + hour
        if self.minute < 10:
            minute = "0" + minute

        return hour + ":" + minute


class Repetition:

    def __init__(self, times, date, current):
        self.times = times
        self.date = date
        self.current = current

    def __eq__(self, other):
        return self.times == other.times and self.date == other.date and self.current == other.current

    def __str__(self):
        return "  Times: " + str(self.times) + "\n  Date: " + str(self.date) + "\n  Current: " + str(self.current)


class Event:

    def __init__(self, pin, days, time, repetition, duration):
        self.pin = pin
        self.days = days
        self.time = time
        self.repetition = repetition
        self.duration = duration

    def __eq__(self, other):
        return self.pin == other.pin and self.days == other.days and self.time == other.time and self.repetition == other.repetition and self.duration == other.duration

    def __str__(self):
        return "Pin: " + str(self.pin) + "\nDays: " + str(self.days) + "\nTime: " + str(self.time) + "\nRepetition: " + str(self.repetition) + "\nDuration: " + str(self.duration)

File: api/models/event.go
package models

import (
	"context"
	"net/http"
	"time"

	"github.com/Loptt/home-automation-system/api/db"
	"github.com/Loptt/home-automation-system/api/errors"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
)

type eventTime struct {
	Hour   int `bson:"hour,omitempty" json:"hour"`
	Minute int `bson:"minute,omitempty" json:"minute"`
}

// eventRepetition represents how an event should be repeated. If times is greater than 0,
// the event will be repeated N times before being deleted. If date is anything later than
// Unix epoch, the event will be executed as long as the current date is less that the
// specified date.
// If both are set to their default values, then the event will be executed indefinitely.
type eventRepetition struct {
	Times   int       `bson:"times,omitempty" json:"times"`
	Date    time.Time `bson:"date,omitempty" json:"date"`
	Current int       `bson:"current,omitempty" json:"current"`
}

// Event represents a trigger event for a device
type Event struct {
	ID         primitive.ObjectID `bson:"_id,omitempty" json:"_id"`
	Name       string             `bson:"name,omitempty" json:"name"`
	Days       []int              `bson:"days,omitempty" json:"days"`
	Time       eventTime          `bson:"time,omitempty" json:"time"`
	Duration   int                `bson:"duration,omitempty" json:"duration"`
	Repetition eventRepetition    `bson:"repetition,omitempty" json:"repetition"`
	Device     primitive.ObjectID `bson:"device,omitempty" json:"device"`
}

// EventController interacts with the event collection in the DB.
type EventController struct {
	collection *mongo.Collection
}

// GetAll retrieves all events.
func (ec *EventController) GetAll(ctx context.Context) ([]Event, error) {
	cursor, err := ec.collection.Find(ctx, bson.D{})
	if err != nil {
		return nil, err
	}

	events := make([]Event, 0)

	for cursor.Next(ctx) {
		var event Event
		if err := cursor.Decode(&event); err != nil {
			return nil, err
		}
		events = append(events, event)
	}

	return events, nil
}

// GetByID retrieves a device given a MongoDB ID.
func (ec *EventController) GetByID(ctx context.Context, id primitive.ObjectID) (*Event, error) {
	var event Event
	eventResult := ec.collection.FindOne(ctx, bson.M{"_id": id})
	if eventResult.Err() != nil {
		if eventResult.Err() == mongo.ErrNoDocuments {
			return nil, errors.NewServerError("No events found", http.StatusNotFound)
		}
		return nil, eventResult.Err()
	}

	eventResult.Decode(&event)

	return &event, nil
}

// GetByDevice retrieves all events for a device
func (ec *EventController) GetByDevice(ctx context.Context, device primitive.ObjectID) ([]Event, error) {
	cursor, err := ec.collection.Find(ctx, bson.M{"device": device})
	if err != nil {
		return nil, err
	}

	events := make([]Event, 0)

	for cursor.Next(ctx) {
		var event Event
		if err := cursor.Decode(&event); err != nil {
			return nil, err
		}
		events = append(events, event)
	}

	return events, nil
}

// Create a new event in the db.
func (ec *EventController) Create(ctx context.Context, event *Event) error {
	_, err := ec.collection.InsertOne(ctx, *event)
	if err != nil {
		return err
	}

	return nil
}

// NewEventController creates a DeviceController with the correct collection.
func NewEventController() (*EventController, error) {
	db, err := db.Database()
	if err != nil {
		return nil, err
	}
	return &EventController{collection: db.Collection("events")}, nil
}
